crates/astria-sequencer/src/transaction/fees.rs

```diff
@@ -0,0 +1,683 @@
+use std::collections::HashMap;
+
+use astria_core::{
+    primitive::v1::{
+        asset::{
+            self,
+            IbcPrefixed,
+        },
+        RollupId,
+    },
+    protocol::{
+        genesis::v1alpha1::Fees,
+        transaction::v1alpha1::{
+            action::{
+                Action,
+                BridgeLockAction,
+                FeeAssetChangeAction,
+                FeeChange,
+                FeeChangeAction,
+            },
+            UnsignedTransaction,
+        },
+    },
+    sequencerblock::v1alpha1::block::Deposit,
+};
+use astria_eyre::eyre::{
+    eyre,
+    OptionExt as _,
+    Result,
+    WrapErr as _,
+};
+use cnidarium::{
+    StateRead,
+    StateWrite,
+};
+use tendermint::abci::{
+    Event,
+    EventAttributeIndexExt as _,
+};
+use tracing::{
+    instrument,
+    warn,
+};
+
+use crate::{
+    accounts::{
+        AddressBytes,
+        StateReadExt as _,
+        StateWriteExt,
+    },
+    assets::StateReadExt as _,
+    authority::StateReadExt,
+    bridge::StateReadExt as _,
+    ibc::{
+        ics20_withdrawal::establish_withdrawal_target,
+        StateReadExt as _,
+    },
+    sequence::StateReadExt as _,
+    transaction::{
+        checks::PaymentMap,
+        StateReadExt as _,
+    },
+};
+
+#[derive(Debug)]
+pub(crate) struct FeeInfo {
+    pub(crate) amt: u128,
+    pub(crate) events: Vec<Event>,
+}
+
+struct PaymentInfo<'a> {
+    from: Option<[u8; 20]>,
+    to: Option<[u8; 20]>,
+    allowed_fee_assets: &'a mut Vec<IbcPrefixed>,
+    fee_asset: asset::Denom,
+    fees_by_asset: &'a mut HashMap<asset::IbcPrefixed, u128>,
+    fee_payment_map: &'a mut HashMap<([u8; 20], [u8; 20], asset::Denom), FeeInfo>,
```

Comment:
make a struct for `([u8; 20], [u8; 20], asset::Denom)` so the fields can be named:
```
struct Payment {
    from: [u8; 20],
    to: [u8; 20],
    denom: asset::Denom,
}
```

Reply:
Switched to struct use in 7c1e05c.